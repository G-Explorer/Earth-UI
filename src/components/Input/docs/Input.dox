@summary > Input is a controlled component, it **always shows React binding value**. Under normal circumstances, input event should be handled. Its handler should update component's binding value. Otherwise, input box's value will not change.

/**
 * @title Basic usage
 * @desc The Input default width is `100%`, you can use `with` property to customize input (px) or use the layout components: `Row`, `Col`.
 */
import { Input } from 'earth-ui'
import { Row, Col } from 'earth-ui'

const InputNormal = () => {
  return (
    <div>
      <Input placeholder="Please input text here" width={456} />
      <br/>
      <Row>
        <Col col="md-14">
          <Input placeholder="Please input text here" />
        </Col>
      </Row>
    </div>
  )
}

/**
 * @title Readonly / Disabled
 * @desc Disable the Input with the `disabled` attribute, but you can also choose `readonly` style for a better experience.
 */
import { Input } from 'earth-ui'

const InputRD = () => {
  return (
    <div>
      <Input value="Readonly text" width={456} readonly />
      <br/>
      <Input value="Disabled Text" width={456} disabled />
    </div>
  )
}

/**
 * @title Icon
 * @desc To add icons in Input, you can simply use `prefix` and `suffix` render-props which can make nested components.
 */
import { Input, Icon, Tooltip } from 'earth-ui'

const InputIcon = () => {
  return (
    <div>
      <Input prefix={<Icon type="search"/>} placeholder="Please input text" width={456} />
      <br/>
      <Input
        suffix={<Tooltip direction="up" title="this is hint"><Icon type="question" /></Tooltip>}
        placeholder="Please input text" width={456}
      />
    </div>
  )
}

/**
 * @title Size
 * @desc Add `size` attribute to change the size of Input. In addition to the default `md` size, there are three other options: `sm` (30px), `md` (35px) and `lg` (40px).
 */
import { Input } from 'earth-ui'

const InputSize = () => {
  return (
    <div>
      <Input suffix={<Icon type="calendar"/>} placeholder="Small" size="sm" width={456} />
      <br/>
      <Input suffix={<Icon type="calendar"/>} placeholder="Medium(default)" width={456} />
      <br/>
      <Input suffix={<Icon type="calendar"/>} placeholder="Large" size="lg" width={456} />
    </div>
  )
}

/**
 * @title Clearable
 * @desc Make the Input `clearable` with the clearable attribute.
 */
import { Input } from 'earth-ui'

const InputClearable = () => {
  return (
    <div>
      <Input placeholder="Please input text here" width={456} clearable/>
    </div>
  )
}

@component Input
