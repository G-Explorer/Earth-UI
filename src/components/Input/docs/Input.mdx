---
props: Input
---

> Input is a controlled component and **does not support** `defaultValue` which is uncontrolled, it **always shows React binding value**. Under normal circumstances, input `onChange` should be handled. Its handler should update component's binding value. Otherwise, input box's value will not change.

## Basic usage

The Input default width is `100%`, you can use `with` property to customize input (px) or use the layout components: `Row`, `Col`.

```jsx tabs
import { Input } from '@webapps-ui/core-react'
import { Row, Col } from '@webapps-ui/core-react'
import { useState } from 'react'

const InputNormal = () => {
  const [value, setValue] = useState('')
  const handleChange = e => {
    setValue(e.target.value)
  }
  return (
    <div>
      <Input value={value} onChange={handleChange} placeholder="Please input text here" width={456} />
      <br/>
      <Row>
        <Col col="md-14">
          <Input value={value} onChange={handleChange} placeholder="Please input text here" />
        </Col>
      </Row>
    </div>
  )
}
```

## Readonly / Disabled

Disable the Input with the `disabled` attribute, but you can also choose `readonly` style for a better experience.

```jsx tabs
import { Input } from '@webapps-ui/core-react'

const InputRD = () => {
  return (
    <div>
      <Input value="Readonly text" width={456} readonly />
      <br/>
      <Input value="Disabled Text" width={456} disabled />
    </div>
  )
}
```

## Icon

To add icons in Input, you can simply use `prefix` and `suffix` render-props which can make nested components.

```jsx tabs
import { Input, Icon, Tooltip } from '@webapps-ui/core-react'
import { useState } from 'react'

const InputIcon = () => {
  const [value, setValue] = useState('')
  const handleChange = e => {
    setValue(e.target.value)
  }
  return (
    <div>
      <Input
        prefix={<Icon type="search"/>}
        value={value}
        onChange={handleChange}
        placeholder="Please input text"
        width={456}
      />
      <br/>
      <Input
        suffix={<Tooltip direction="up" title="this is hint"><Icon type="question" /></Tooltip>}
        value={value}
        onChange={handleChange}
        placeholder="Please input text"
        width={456}
      />
    </div>
  )
}
```

## Size

Add `size` attribute to change the size of Input. In addition to the default `md` size, there are three other options: `sm` (30px), `md` (35px) and `lg` (40px).

```jsx tabs
import { Input } from '@webapps-ui/core-react'
import { useState } from 'react'

const InputSize = () => {
  const [value, setValue] = useState('')
  const handleChange = e => {
    setValue(e.target.value)
  }
  return (
    <div>
      <Input size="sm" value={value} onChange={handleChange} suffix={<Icon type="calendar"/>} placeholder="Small" width={456} />
      <br/>
      <Input size="md" value={value} onChange={handleChange} suffix={<Icon type="calendar"/>} placeholder="Medium(default)" width={456} />
      <br/>
      <Input size="lg" value={value} onChange={handleChange} suffix={<Icon type="calendar"/>} placeholder="Large" width={456} />
    </div>
  )
}
```

## Clear

Triggers when the Input is cleared by clicking the clear button

```jsx tabs
import { Input } from '@webapps-ui/core-react'
import { useState } from 'react'

const InputClearable = () => {
  const [value, setValue] = useState('')
  const handleChange = e => {
    setValue(e.target.value)
  }
  const handleClear = () => {
    setValue('')
  }
  return (
    <Input
      onClear={handleClear}
      value={value}
      onChange={handleChange}
      placeholder="Please input text"
      width={456}
    />
  )
}
```
