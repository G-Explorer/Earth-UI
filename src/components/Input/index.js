import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import Icon from '../Icon'
import omit from '../_utils/omit'
import './index.less'

function fixControlledValue (value) {
  return value || ''
}

class Input extends React.Component {
  constructor (props) {
    super(props)
    this.input = React.createRef()
    this.state = {
      focus: false,
      mouseEnter: false
    }
  }

  handleChange = e => {
    this.props.onChange && this.props.onChange(e)
  }

  handleClear = e => {
    if (this.props.onClear) {
      this.props.onClear(e)
      this.focus()
    }
  }

  handleFocus = v => {
    this.setState({ focus: v })
  }

  handleMouseEnter = v => {
    this.setState({ mouseEnter: v })
  }

  /**
   * @public
   * @name ref.current.focus
   * @description Same as HTMLInputElement.focus()
   */
  focus () {
    this.input.current.focus()
  }

  /**
   * @public
   * @name ref.current.blur
   * @description Same as HTMLInputElement.blur()
   */
  blur () {
    this.input.current.blur()
  }

  /**
   * @public
   * @name ref.current.select
   * @description Same as HTMLInputElement.select()
   */
  select () {
    this.input.current.select()
  }

  render () {
    const {
      value,
      className,
      size,
      width,
      prefix,
      suffix,
      readonly,
      onClear,
      ...other
    } = this.props

    const otherProps = omit(other, ['defaultValue', 'onChange'])

    const classNames = cx(
      `${prefixCls}-input`,
      {
        [`${prefixCls}-input--${size}`]: size,
        [`${prefixCls}-input--prefix`]: prefix,
        [`${prefixCls}-input--suffix`]: suffix || readonly || onClear
      },
      className
    )
    if (width) {
      otherProps.style = Object.assign(otherProps.style || {}, { width })
    }
    if (prefix || suffix) {
      return (
        <div className={`${prefixCls}-input__affix-wrapper`} style={{ width }}>
          {prefix && (
            <span className={`${prefixCls}-input__affix-wrapper--prefix`}>
              {prefix}
            </span>
          )}
          <input
            {...otherProps}
            className={classNames}
            value={fixControlledValue(value)}
            onChange={this.handleChange}
            ref={this.input}
          />
          {suffix && (
            <span className={`${prefixCls}-input__affix-wrapper--suffix`}>
              {suffix}
            </span>
          )}
        </div>
      )
    }
    if (readonly) {
      return (
        <div className={`${prefixCls}-input__affix-wrapper`} style={{ width }}>
          <input
            {...otherProps}
            className={classNames}
            readOnly={readonly}
            value={fixControlledValue(value)}
            onChange={this.handleChange}
            ref={this.input}
          />
          <span className={`${prefixCls}-input__affix-wrapper--suffix`}>
            <Icon type="locked" />
          </span>
        </div>
      )
    }
    if (onClear) {
      const { focus, mouseEnter } = this.state
      return (
        <div
          className={`${prefixCls}-input__affix-wrapper`}
          style={{ width }}
          onMouseEnter={() => this.handleMouseEnter(true)}
          onMouseLeave={() => this.handleMouseEnter(false)}
        >
          <input
            {...otherProps}
            className={classNames}
            value={fixControlledValue(value)}
            onChange={this.handleChange}
            onBlur={() => this.handleFocus(false)}
            onFocus={() => this.handleFocus(true)}
            ref={this.input}
          />
          {value && (focus || mouseEnter) && !otherProps.disabled && (
            <span className={`${prefixCls}-input__affix-wrapper--suffix`}>
              <Icon type="remove" onClick={this.handleClear} />
            </span>
          )}
        </div>
      )
    }
    return (
      <input
        {...otherProps}
        className={classNames}
        value={fixControlledValue(value)}
        onChange={this.handleChange}
        ref={this.input}
      />
    )
  }
}

Input.propTypes = {
  className: PropTypes.string,

  /** binding value */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /** riggers when the icon inside Input value change */
  onChange: PropTypes.func,

  /** type of input, support [native input types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#Form_%3Cinput%3E_types) */
  type: PropTypes.string,

  /** size of Input */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),

  /** the length of input, unit as px */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

  /** whether Input is disabled */
  disabled: PropTypes.bool,

  /** whether Input is readonly */
  readonly: PropTypes.bool,

  /** same as native input placeholder */
  placeholder: PropTypes.string,

  /** the prefix icon for the Input */
  prefix: PropTypes.element,

  /** the suffix icon for the Input */
  suffix: PropTypes.element,

  /** triggers when the Input is cleared by clicking the clear button */
  onClear: PropTypes.func
}

Input.defaultProps = {
  width: '100%',
  size: 'md',
  type: 'text'
}

export default Input
