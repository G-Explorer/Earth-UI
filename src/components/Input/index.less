@import '~ui-variables';
@import '~ui-mixins';

@width: 100%;

.@{prefix-cls}-input:not([type='checkbox']):not([type='radio']) {
  border: 1px solid @ui-color-gray-light;
  color: @ui-color-black;
  cursor: text;
  display: block;
  width: 100%;
  font-weight: @ui-fontweight;
  background-color: @ui-color-white;
  white-space: nowrap;
  word-wrap: normal;
  overflow: hidden;
  text-overflow: ellipsis;
  transition: border-color @ui-transition-fast ease;
  &[disabled] {
    background-color: fade(@ui-color-gray-lightest, 60%);
    color: @ui-color-gray;
    cursor: not-allowed;
  }
  &[readonly] {
    color: @ui-color-gray;
    cursor: default;
  }
}

// Size ........................................................................

.@{prefix-cls}-input--sm:not([type='checkbox']):not([type='radio']) {
  height: @ui-unit-plus;
  font-size: @ui-fontsize-mini;
  padding-left: @ui-unit-half;
  padding-right: @ui-unit-half;
  border-radius: @ui-radius;
}

.@{prefix-cls}-input--md:not([type='checkbox']):not([type='radio']) {
  height: @ui-unit-plus + @ui-unit-quarter;
  font-size: @ui-fontsize;
  padding-left: @ui-unit-half + 2px;
  padding-right: @ui-unit-half + 2px;
  border-radius: @ui-radius-md;
}

.@{prefix-cls}-input--lg:not([type='checkbox']):not([type='radio']) {
  height: @ui-unit-double;
  font-size: @ui-fontsize;
  padding-left: @ui-unit-half + @ui-unit-quarter;
  padding-right: @ui-unit-half + @ui-unit-quarter;
  border-radius: @ui-radius-md;
}

// Prefix & Suffix ..........................................................

.@{prefix-cls}-input__affix-wrapper {
  position: relative;
  &--prefix, &--suffix {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    i:before {
      .ui-mixin-icon(@ui-fontsize-medium);
      width: @ui-unit-plus;
      color: @ui-color-slate-lightest;
    }
  }
  &--prefix {
    left: 12px;
  }
  &--suffix {
    right: 12px;
  }
}

.@{prefix-cls}-input--prefix:not([type='checkbox']):not([type='radio']) {
  padding-left: @ui-unit-plus;
}

.@{prefix-cls}-input--suffix:not([type='checkbox']):not([type='radio']) {
  padding-right: @ui-unit-plus;
}

// Focus & Hover ...............................................................

.local-mixin-focus-outline(@color: @ui-color-blue) {
  border: 1px solid @color;
  box-shadow: 0 0 0 1px fade(@color, 75%);
  border-radius: @ui-radius;
}
.local-mixin-hover-outline(@color: @ui-color-blue) {
  .local-mixin-focus-outline(fade(@color, 25%));
}
.local-mixin-no-outline() {
  border: 1px solid transparent !important;
  box-shadow: none !important;
}

.@{prefix-cls}-input[type="text"]:not(:disabled):not(:read-only) {
  &:hover {
    .local-mixin-hover-outline();
  }
  &:focus {
    .local-mixin-focus-outline();
  }
}
