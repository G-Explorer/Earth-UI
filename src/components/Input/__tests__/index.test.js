import React from 'react'
import { mount } from 'enzyme'
import Input from '..'
import focusTest from '../../../tests/shared/focusTest'
import mountTest from '../../../tests/shared/mountTest'

describe('Input basic usage', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {})

  afterEach(() => {
    errorSpy.mockReset()
  })

  afterAll(() => {
    errorSpy.mockRestore()
  })

  focusTest(Input)
  mountTest(Input)

  it('should have no error when value prop without onChange handler', () => {
    // value means default value in this case
    mount(<Input value="111" />)
    expect(errorSpy).not.toHaveBeenCalled()
  })

  it('should set width to 100% as default', () => {
    const wrapper = mount(<Input value="111" />) // should
    expect(wrapper.find('input').prop('style').width).toBe('100%')
    expect(wrapper).toMatchSnapshot()
  })

  it('should support specific px width', () => {
    const wrapper = mount(<Input width={456} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should set empty string if value is undefined, null', () => {
    const wrappers = [null, undefined].map(val => mount(<Input value={val} />))
    wrappers.forEach(wrapper => {
      expect(wrapper.find('input').getDOMNode().value).toEqual('')
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('should change value when input other value', () => {
    const state = { value: '111' }
    const handleChange = e => {
      wrapper.setProps({ value: e.target.value })
    }
    const wrapper = mount(<Input value={state.value} onChange={handleChange} />)
    wrapper.find('input').simulate('change', { target: { value: '222' } })
    expect(wrapper.find('input').getDOMNode().value).toEqual('222')
    expect(wrapper).toMatchSnapshot()
  })
})

// todo
// describe('Input Readonly / Disabled', () => {})
// describe('Input Icon', () => {})
// describe('Input Size', () => {})

describe('Input clear', () => {
  const state = { value: '111' }
  const handleClear = jest.fn()

  it('should hide icon when input have not focus and mouseenter', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    expect(wrapper.find('input').getDOMNode().value).toEqual('111')
    expect(wrapper.find('.true-icon').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show icon correctly when input focus', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('input').simulate('focus')
    expect(wrapper.find('input').getDOMNode().value).toEqual('111')
    expect(wrapper.find('.true-icon').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should hide icon correctly when input blur', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('input').simulate('focus')
    wrapper.find('input').simulate('blur')
    expect(wrapper.find('input').getDOMNode().value).toEqual('111')
    expect(wrapper.find('.true-icon').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show icon correctly when mouse enter Input boundary', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('div').simulate('mouseenter')
    expect(wrapper.find('input').getDOMNode().value).toEqual('111')
    expect(wrapper.find('.true-icon').exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show icon correctly when mouse leave Input boundary', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('div').simulate('mouseenter')
    wrapper.find('div').simulate('mouseleave')
    expect(wrapper.find('input').getDOMNode().value).toEqual('111')
    expect(wrapper.find('.true-icon').exists()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should not show icon if value is undefined, null or empty string', () => {
    const wrappers = [null, undefined, ''].map(val =>
      mount(<Input value={val} onClear={handleClear} />)
    )
    wrappers.forEach(wrapper => {
      expect(wrapper.find('input').getDOMNode().value).toEqual('')
      expect(wrapper.find('.true-icon').exists()).toEqual(false)
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('should clear value correctly when click clear icon', () => {
    let argumentEventObject
    const handleClear = e => {
      argumentEventObject = e
      wrapper.setProps({ value: '' })
    }
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('input').simulate('focus')
    wrapper
      .find('.true-icon')
      .at(0)
      .simulate('click')
    expect(argumentEventObject.type).toBe('click')
    expect(wrapper.find('input').getDOMNode().value).toEqual('')
    expect(wrapper).toMatchSnapshot()
  })

  it('should focus input after clear', () => {
    const wrapper = mount(<Input value={state.value} onClear={handleClear} />)
    wrapper.find('input').simulate('focus')
    wrapper
      .find('.true-icon')
      .at(0)
      .simulate('click')
    expect(document.activeElement).toBe(
      wrapper
        .find('input')
        .at(0)
        .getDOMNode()
    )
  })

  it('should not support clearable when it is disabled', () => {
    const wrapper = mount(
      <Input value={state.value} onClear={handleClear} disabled />
    )
    expect(wrapper.find('.true-icon').length).toBe(0)
  })
})
