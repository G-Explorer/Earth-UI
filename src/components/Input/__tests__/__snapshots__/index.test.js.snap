// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input basic usage should change value when input other value 1`] = `
<Input
  onChange={[Function]}
  size="md"
  type="text"
  value="222"
  width="100%"
>
  <input
    className="true-input true-input--md"
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text"
    value="222"
  />
</Input>
`;

exports[`Input basic usage should set empty string if value is undefined, null 1`] = `
<Input
  size="md"
  type="text"
  value={null}
  width="100%"
>
  <input
    className="true-input true-input--md"
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text"
    value=""
  />
</Input>
`;

exports[`Input basic usage should set empty string if value is undefined, null 2`] = `
<Input
  size="md"
  type="text"
  width="100%"
>
  <input
    className="true-input true-input--md"
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text"
    value=""
  />
</Input>
`;

exports[`Input basic usage should set width to 100% as default 1`] = `
<Input
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <input
    className="true-input true-input--md"
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text"
    value="111"
  />
</Input>
`;

exports[`Input basic usage should support specific px width 1`] = `
<Input
  size="md"
  type="text"
  width={456}
>
  <input
    className="true-input true-input--md"
    onChange={[Function]}
    style={
      Object {
        "width": 456,
      }
    }
    type="text"
    value=""
  />
</Input>
`;

exports[`Input clear should clear value correctly when click clear icon 1`] = `
<Input
  onClear={[Function]}
  size="md"
  type="text"
  value=""
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value=""
    />
  </div>
</Input>
`;

exports[`Input clear should hide icon correctly when input blur 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
  </div>
</Input>
`;

exports[`Input clear should hide icon when input have not focus and mouseenter 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
  </div>
</Input>
`;

exports[`Input clear should not show icon if value is undefined, null or empty string 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value={null}
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value=""
    />
  </div>
</Input>
`;

exports[`Input clear should not show icon if value is undefined, null or empty string 2`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value=""
    />
  </div>
</Input>
`;

exports[`Input clear should not show icon if value is undefined, null or empty string 3`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value=""
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value=""
    />
  </div>
</Input>
`;

exports[`Input clear should show icon correctly when input focus 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
    <span
      className="true-input__affix-wrapper--suffix"
    >
      <Icon
        onClick={[Function]}
        type="remove"
      >
        <i
          className="true-icon true-icon__remove"
          onClick={[Function]}
        />
      </Icon>
    </span>
  </div>
</Input>
`;

exports[`Input clear should show icon correctly when mouse enter Input boundary 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
    <span
      className="true-input__affix-wrapper--suffix"
    >
      <Icon
        onClick={[Function]}
        type="remove"
      >
        <i
          className="true-icon true-icon__remove"
          onClick={[Function]}
        />
      </Icon>
    </span>
  </div>
</Input>
`;

exports[`Input clear should show icon correctly when mouse leave Input boundary 1`] = `
<Input
  onClear={[MockFunction]}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
  </div>
</Input>
`;

exports[`Input disabled should have disabled prop on native input 1`] = `
<Input
  disabled={true}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <input
    className="true-input true-input--md"
    disabled={true}
    onChange={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
    type="text"
    value="111"
  />
</Input>
`;

exports[`Input icon should show prefix icon correctly when have prefix render props 1`] = `
<Input
  prefix={
    <Icon
      type="search"
    />
  }
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <span
      className="true-input__affix-wrapper--prefix"
    >
      <Icon
        type="search"
      >
        <i
          className="true-icon true-icon__search"
        />
      </Icon>
    </span>
    <input
      className="true-input true-input--md true-input--prefix"
      onChange={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
  </div>
</Input>
`;

exports[`Input icon should show suffix icon correctly when have suffix render props 1`] = `
<Input
  size="md"
  suffix={
    <Tooltip
      align="middle"
      direction="up"
      title="this is hint"
      triggerMode="hover"
    >
      <Icon
        type="question"
      />
    </Tooltip>
  }
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onChange={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
    <span
      className="true-input__affix-wrapper--suffix"
    >
      <Tooltip
        align="middle"
        direction="up"
        title="this is hint"
        triggerMode="hover"
      >
        <Popover
          align="middle"
          className="true-tooltip__popover"
          content="this is hint"
          direction="up"
          triggerMode="hover"
        >
          <Icon
            className="true-popover__trigger"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="question"
          >
            <i
              className="true-icon true-icon__question true-popover__trigger"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            />
          </Icon>
        </Popover>
      </Tooltip>
    </span>
  </div>
</Input>
`;

exports[`Input readonly should have readonly prop on native input 1`] = `
<Input
  readonly={true}
  size="md"
  type="text"
  value="111"
  width="100%"
>
  <div
    className="true-input__affix-wrapper"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <input
      className="true-input true-input--md true-input--suffix"
      onChange={[Function]}
      readOnly={true}
      style={
        Object {
          "width": "100%",
        }
      }
      type="text"
      value="111"
    />
    <span
      className="true-input__affix-wrapper--suffix"
    >
      <Icon
        type="locked"
      >
        <i
          className="true-icon true-icon__locked"
        />
      </Icon>
    </span>
  </div>
</Input>
`;
